onnxruntime/core/providers/cpu/math/det.cc

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/math/det.h"
+#include "core/util/math_cpuonly.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+ONNX_CPU_OPERATOR_KERNEL(
+    Det,
+    11,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
+    Det<float>);
+
+template <typename T>
+Status Det<T>::Compute(OpKernelContext* context) const {
+  const auto* X = context->Input<Tensor>(0);
+  ORT_ENFORCE(X != nullptr);
+
+  const auto& X_shape = X->Shape();
+  int X_num_dims = static_cast<int>(X_shape.NumDimensions());
+
+  // input validation
+  if (X_num_dims < 2) {  // this is getting capture by shape inference code as well
+    return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Input tensor should have a rank of at least 2");
+  }
+  if (X_shape[X_num_dims - 1] != X_shape[X_num_dims - 2]) {  // this is getting capture by shape inference code as well
+    return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Matrix dimensions are not equal. Square matrix is expected");
+  }
+
+  const auto* X_data = X->Data<T>();
+  int matrix_dim = static_cast<int>(X_shape[X_num_dims - 1]);
+
+  auto get_determinant = [matrix_dim](const T* matrix_ptr) -> T {
+    auto one_eigen_mat = ConstEigenMatrixMapRowMajor<T>(matrix_ptr, matrix_dim, matrix_dim);
+    return one_eigen_mat.determinant();
+  };
+
+  if (X_num_dims == 2) {
+    auto* Y = context->Output(0, {});  // as per spec output should be a scalar when input is 2D
+    auto* Y_data = Y->template MutableData<T>();
+    *Y_data = get_determinant(X_data);
+  } else {
+    int batch_size = static_cast<int>(X_shape[0]);
```

Comment:
Per spec the number of batch dimensions can be greater than 1.
> Det takes one input tensor of shape [*, M, M], where * is zero or more batch dimensions

Reply:
can you give an example?

Reply:
Yes I believe Bowen is right and he gives an example in his comment below.

I think there is only a slight change needed in the existing logic to accommodate that. 

Instead of batch_size, there can be a variable 'num_matrices' and that value can be arrived at by multiplying the outermost dimensions (except the innermost 2 which are matrix dimensions). The code just has to iterate over num_matrices which for practical purposes is the batch_size.

When input is just 2D, num_matrices = 1.

Reply:
fixed